## CNV calling for HCT116 NOMe-seq
library(biscuiteer)
library(Homo.sapiens) 
library(QDNAseq.hg19)
library(future)

#load in the joint HCT116 and DKO NOMe-seq hcg data
#mergecg was *not* run on this data
#data location
#setwd("/secondary/projects/triche/ben_projects/biscuit_manuscript/analysis")
hct116_hcg <- readBiscuit(BEDfile = "joint_nome_hct116_hcg.bed.gz",
                          VCFfile = "joint_nome_hct116.vcf.gz",
                          genome = "hg19",
                          merged = FALSE)

#save it
saveRDS(hct116_hcg, file = "hct116_hcg.bsseq.rds")

#bin the coverage
bins <- getBinAnnotations(binSize=30, genome="hg19") # 30kb
#Loaded bin annotations for genome ‘hg19’, bin size 30 kbp, and experiment type ‘SR50’ from annotation package QDNAseq.hg19 v1.16.0
whichChroms <- subset(as(seqinfo(Homo.sapiens), "GRanges"),
                      seqnames %in% paste0("chr", 1:22))

hct116_hcg_bincov <- binCoverage(hct116_hcg, bins, which=whichChroms)
# Converting QDNAseq bins to GRanges for coverage
# Warning message:
#   In `seqlevelsStyle<-`(`*tmp*`, value = value) :
#   more than one seqlevels style supplied, using the 1st one only


show(hct11hct116_hcg_bincov)
# QDNAseqReadCounts (storageMode: lockedEnvironment)
# assayData: 96043 features, 2 samples
# element names: counts
# protocolData: none
# phenoData
# sampleNames: SRR1534687.sorted.markdup SRR1534707.sorted.markdup
# varLabels: name total.reads used.reads expected.variance
# varMetadata: labelDescription
# featureData
# featureNames: 1:1-30000 1:30001-60000 ... 22:51300001-51304566 (96043
#                                                                 total)
# fvarLabels: chromosome start ... use (9 total)
# fvarMetadata: labelDescription
# experimentData: use 'experimentData(object)'
# Annotation: generated by biscuiteer 1.0.0

#saveit
saveRDS(hct116_hcg_bincov, file = "hct116_hcg_bincov.hg19.rds")

#bin and segment
hct116_hcg_bincov_filt <- applyFilters(hct116_hcg_bincov,
                                       residual=TRUE,
                                       blacklist=TRUE)
# 96,043	total bins
# 96,043	of which in selected chromosomes
# 89,705	of which with reference sequence
# 83,607	final bins
hct116_hcg_bincov_filt <- estimateCorrection(hct116_hcg_bincov_filt)

#saveit
saveRDS(hct116_hcg_bincov_filt,
        file = "hct116_hcg_bincov_filt_30kb.rds")

CNVs <- correctBins(hct116_hcg_bincov_filt)
#Note: Filtering out additional 23 bins due to missing values.

CNVsNormalized <- normalizeBins(CNVs)
CNVsSmooth <- smoothOutlierBins(CNVsNormalized)
# this is fast because it just uses DNAcopy for CBS
CNVsSegmented <- segmentBins(CNVsSmooth)
# Not sure what is going on with the NA label messages
CNVsSegmented <- normalizeSegmentedBins(CNVsSegmented)

#call bins
CNVsCallBins <- callBins(CNVsSegmented)

# EM algorithm started ...
# 
# [1] "Total number of segments present in the data: 609"
# [1] "Number of segments used for fitting the model: 200"
# 346482823150063231850.52403.41093525318432117045840.16433.2480878984350368852568.23319.1
# Calling iteration1:
#   
#   optim results
# time: 8
# minimum: 32847.3508803229
# 232782.9069804116-0.920848166126444-0.3277653001998180.0110817430462270.2943790597607130.5032443265983630.9361503091740310.4427321419993820.1015441914690580.09980120983859270.1147005013617120.1424375259233440.143975035300216
# EM algorithm done ...
# 
# Computing posterior probabilities for all segments ...
# 
# Total time:1minutes
# 
# Adjusting segmented data for cellularity ...
# 
# Cellularity sample1: 1
# 
# Cellularity sample2: 1
# 
# Adjusting normalized data for cellularity ...
# 
# Cellularity sample1: 1
# 
# Cellularity sample2: 1
# 
# 1
# 346523483154214591850.72406.51093525318432117045840.16433.2999807154350368855339.63319.1
# 346523663158393871850.72409.71093525318432117045840.16433.2999807154366291145339.63331.3
# 346523643158393841850.72409.71093525318432117045840.16433.2999807154366291145339.63331.3
# 346523913165100871850.72414.81093525318432117045840.16433.2999807154366291145339.63331.3
# 346528183165110451850.72414.81093525318432117045840.16433.2999807154366291145339.63331.3
# 346528273165110491850.72414.81093525318432117045840.16433.2999807154366291145339.63331.3
# 346528363165110531850.72414.81093525318432117045840.16433.2999807154366291145339.63331.3
# 346528453165110571850.72414.81093525318432117045840.16433.2999807154366291145339.63331.3
# 346528543165110611850.72414.81093525318432117045840.16433.2999807154366291145339.63331.3
# 346528603165110641850.72414.81093525318432117045840.16433.2999807154366291145339.63331.3
# 346528813165111021850.72414.81093525318432117045840.16433.2999807154366291145339.63331.3
# 346533053165114291850.72414.81093525318432117045840.16433.2999807154366291145339.63331.3
# 2
# 346533233172636951850.72420.61093525318432117045840.16433.2999807154366291145339.63331.3
# 346533253172636981850.72420.61093525318432117045840.16433.2999807154403091155339.63359.3
# 346533233172636951850.72420.61093525318432117045840.16433.2999807154403091155339.63359.3
# 346533263173475321850.72421.21093525318432117045840.16433.2999807154403091155339.63359.3
# 346533263173475321850.72421.21093525318432117045840.16433.2999807154403091155339.63359.3
# 346533353173475361850.72421.21093525318432117045840.16433.2999807154403091155339.63359.3
# 346533443173475401850.72421.21093525318432117045840.16433.2999807154403091155339.63359.3
# 346533533173475441850.72421.21093525318432117045840.16433.2999807154403091155339.63359.3
# 346533623173475481850.72421.21093525318432117045840.16433.2999807154403091155339.63359.3
# 346533683173475511850.72421.21093525318432117045840.16433.2999807154403091155339.63359.3
# 346533893173475891850.72421.21093525318432117045840.16433.2999807154403091155339.63359.3
# 3465742031886237018512432.81093525318432117045840.16433.2999807154403091155339.63359.3
# FINISHED!
#   
# Total time:2minutes

#saveit
saveRDS(CNVsCallBins, file = "hct116_dko_called_bins_use_this_with_plotting.rds")

source("~/Documents/Manuscripts/Biscuit_Zhou_Johnson_2019/genomePlot.R")
source("~/Documents/Manuscripts/Biscuit_Zhou_Johnson_2019/QDNAseqToSE.R")
source("~/Documents/Manuscripts/Biscuit_Zhou_Johnson_2019/exportSegments.R")
source("~/Documents/Manuscripts/Biscuit_Zhou_Johnson_2019/checkAssays.R")

#export the seg files
#biscuiteer export of segments... maybe for read.segs?
exportSegments(CNVsCallBins, filename = "hct116_and_dko_cnv_calls_30kb_bins.seg")
#QDNseq export function
exportBins(CNVsCallBins, filename = "hct116_and_dko_cnv_calls_30kb_bins_qdnaseq_exportfunc.seg")

#convert to SE
hct116_cnv_se <- QDNAseqToSE(CNVsCallBins)

#saveit
saveRDS(hct116_cnv_se, file = "hct116_cnv_calls_final_se.rds")

#plot
par(mfcol=c(2,1))
for (i in seq_len(ncol(hct116_cnv_se))) {
  genomePlot(CNVsCallBins[,i], set_par=FALSE)
} # note autoconversion of results 
dev.copy2pdf(file="HCT116_and_DKO_CNV.called.hg19.pdf")


##Well that doesn't look half bad
##Let's confirm with WGS


